### PR TITLE
core: Add AffineMap

### PR DESCRIPTION
**Pull request by Groverkss**
This patch adds AffineMap to core ir. AffineMap allows modeling many dialects in upstream MLIR like Linalg, Affine, Memref (load/store), SCF, etc.

### CODE REVIEW

File: tests/test_affine_builtins.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+
+def test_affine_expr():
+    # d0, s0
+    a = AffineExpr.dimension(0)
+    b = AffineExpr.symbol(0)
+
+    # (5 * d0) + s0 + 1
+    c = (a * 5) + b + 1
+    print(c)
```
Comment (math-fehr) [inline tests/test_affine_builtins.py:12]: For testing here, I would `assert` that this is equal to `AffineExpr` constructed manually (without the operators).

File: xdsl/ir.py [hunk 2]
```diff
@@ -1612,3 +1613,129 @@ def is_structurally_equivalent(
         ):
             return False
         return True
+
+
+class _AffineExprKind(Enum):
```
Comment (math-fehr) [inline xdsl/ir.py:1618]: I would prefer if we would move this outside of this file to be honest!
`ir.py` is already quite big, and it wouldn't hurt to move this a bit away!

File: xdsl/ir.py [hunk 3]
```diff
@@ -1612,3 +1613,129 @@ def is_structurally_equivalent(
...
+    Add = ("+",)
```
Comment (math-fehr) [inline xdsl/ir.py:1619]: Is there a reason you are using tuple values, instead of just `"+"`?

File: xdsl/ir.py [hunk 4]
```diff
@@ -1612,3 +1613,129 @@ def is_structurally_equivalent(
...
+    Mul = ("*",)
+    Mod = ("mod",)
+    FloorDiv = ("floordiv",)
+    CeilDiv = ("ceildiv",)
+    Constant = ("const",)
+    DimId = ("d",)
+    SymbolId = ("s",)
+
+
+class _AffineExprStorage(ABC):
```
Comment (math-fehr) [inline xdsl/ir.py:1629]: Could you write a bit of documentation for each of these classes? Mostly to help us understand.

Comment (math-fehr) [inline xdsl/ir.py:1629]: Also, I would make these class `@dataclasss`, so you generate directly the `__init__`, and the `__eq__`.

File: xdsl/ir.py [hunk 5]
```diff
@@ -1612,3 +1613,129 @@ def is_structurally_equivalent(
...
+    kind: _AffineExprKind
+
+
+class _AffineBinaryOpExprStorage(_AffineExprStorage):
+    lhs: AffineExpr
+    rhs: AffineExpr
+
+    def __init__(self, kind: _AffineExprKind, lhs: AffineExpr, rhs: AffineExpr) -> None:
+        if kind not in [
```
Comment (math-fehr) [inline xdsl/ir.py:1638]: Here I would just put this in a `__post_init__`

File: xdsl/ir.py [hunk 6]
```diff
@@ -1612,3 +1613,129 @@ def is_structurally_equivalent(
...
+            _AffineExprKind.FloorDiv,
+            _AffineExprKind.CeilDiv,
+        ]:
+            raise ValueError(f"Invalid kind {kind} for _AffineBinaryOpExprStorage")
+        self.kind = kind
+        self.lhs = lhs
+        self.rhs = rhs
+
+    def __str__(self) -> str:
+        return f"({str(self.lhs)} {str(self.kind.value[0])} {str(self.rhs)})"
```
Comment (math-fehr) [inline xdsl/ir.py:1651]:
```suggestion
        return f"({self.lhs} {self.kind.value[0]} {self.rhs})"
```

File: xdsl/ir.py [hunk 7]
```diff
@@ -1612,3 +1613,129 @@ def is_structurally_equivalent(
...
+
+
+class _AffineDimExprStorage(_AffineExprStorage):
+    position: int
+
+    def __init__(self, kind: _AffineExprKind, position: int) -> None:
```
Comment (math-fehr) [inline xdsl/ir.py:1657]: Same here with `__post_init__`

File: xdsl/ir.py [hunk 8]
```diff
@@ -1612,3 +1613,129 @@ def is_structurally_equivalent(
...
+        if kind != _AffineExprKind.DimId and kind != _AffineExprKind.SymbolId:
+            raise ValueError(f"Invalid kind {kind} for _AffineDimExprStorage")
+        self.kind = kind
+        self.position = position
+
+    def __str__(self) -> str:
+        return f"{str(self.kind.value[0])}{self.position}"
```
Comment (math-fehr) [inline xdsl/ir.py:1664]:
```suggestion
        return f"{self.kind.value[0]}{self.position}"
```

File: xdsl/ir.py [hunk 9]
```diff
@@ -1612,3 +1613,129 @@ def is_structurally_equivalent(
...
+
+
+class _AffineConstantExprStorage(_AffineExprStorage):
+    value: int
+
+    def __init__(self, value: int) -> None:
```
Comment (math-fehr) [inline xdsl/ir.py:1670]: This one I would just disable the `dataclass` `__init__` and keep that one.

File: xdsl/ir.py [hunk 10]
```diff
@@ -1612,3 +1613,129 @@ def is_structurally_equivalent(
...
+
+    def __sub__(self, other: AffineExpr | int) -> AffineExpr:
+        return (self * -1) + other
+
+    def __mul__(self, other: AffineExpr | int) -> AffineExpr:
+        if isinstance(other, int):
+            other = AffineExpr.constant(other)
+        if other.impl.kind != _AffineExprKind.Constant:
+            # TODO: MLIR also supports multiplication by symbols although I don't like it so I leave this to Mr. Mathew.
+            raise ValueError(
```
Comment (math-fehr) [inline xdsl/ir.py:1711]:
```suggestion
            raise NotImplementedError(
```

File: xdsl/ir.py [hunk 11]
```diff
@@ -1612,3 +1613,129 @@ def is_structurally_equivalent(
...
+        return self * -1
+
+    def __sub__(self, other: AffineExpr | int) -> AffineExpr:
+        return (self * -1) + other
+
+    def __mul__(self, other: AffineExpr | int) -> AffineExpr:
+        if isinstance(other, int):
+            other = AffineExpr.constant(other)
+        if other.impl.kind != _AffineExprKind.Constant:
+            # TODO: MLIR also supports multiplication by symbols although I don't like it so I leave this to Mr. Mathew.
```
Comment (math-fehr) [inline xdsl/ir.py:1710]: I would remove that line ;)

**Review by math-fehr — COMMENTED** [review]
I added some style comments! Otherwise, the code seems good!
I would just like more documentation, and probably some more tests

File: xdsl/ir.py (hunk 2 above)
Comment (webmiche) [reply xdsl/ir.py:1618]: I agree. There is the affine dialect, right? Maybe move it there?

File: tests/test_affine_builtins.py [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+
+
+test_affine_expr()
```
Comment (webmiche) [inline tests/test_affine_builtins.py:15]:
```suggestion
```

**Review by webmiche — COMMENTED** [review]
Thanks for the PR! This stil needs some more tests and some cleaning, but it still is exciting to get this into xDSL. The first step towards xDFPL  :+1:

**Comment by AntonLydike** [comment]
Super happy to see affine being added, thanks @Groverkss! Doing the lords work!

**Comment by tobiasgrosser** [comment]
@Groverkss, this is a great addition. I am looking forward to seeing this upstreamed. This will really allow us to parse one of the most unconventional MLIR dialect. Cool!

File: xdsl/ir.py (hunk 3 above)
Comment (Groverkss) [reply xdsl/ir.py]: Python Enums don't allow that sadly. I replaced it with auto() and a get_token() function instead.

File: xdsl/ir.py [hunk 13]
```diff
@@ -1612,3 +1613,129 @@ def is_structurally_equivalent(
...
+
+class _AffineExprKind(Enum):
+    Add = ("+",)
+    Mul = ("*",)
+    Mod = ("mod",)
+    FloorDiv = ("floordiv",)
+    CeilDiv = ("ceildiv",)
+    Constant = ("const",)
+    DimId = ("d",)
+    SymbolId = ("s",)
```
[the remark this answers is not in the document]
Comment (Groverkss) [reply xdsl/ir.py]: done

File: xdsl/ir.py [hunk 14]
```diff
@@ -1612,3 +1613,235 @@ def is_structurally_equivalent(
...
+                return lhs // rhs
+            elif self._impl.kind == _AffineExprKind.CeilDiv:
+                return -(-lhs // rhs)
+
+        raise ValueError("Unreachable")
+
+    def __add__(self, other: AffineExpr | int) -> AffineExpr:
+        if isinstance(other, int):
+            other = AffineExpr.constant(other)
+        # TODO: Simplify addition here before returning.
```
Comment (superlopuh) [inline xdsl/ir.py:1747]: issue?

**Review by superlopuh — APPROVED** [review]
very cool

**Review by superlopuh — APPROVED** [review]

File: xdsl/affine_ir.py [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+    def __init__(self, value: int) -> None:
+        self.kind = _AffineExprKind.Constant
+        self.value = value
+
+    def __str__(self) -> str:
+        return f"{self.value}"
+
+
+@dataclass()
+class AffineExpr:
```
Comment (math-fehr) [inline xdsl/affine_ir.py:104]: Can you add a bit of documentation on this class? Since this is the class that people will use.

**Review by math-fehr — APPROVED** [review]
Nice!

File: xdsl/affine_ir.py [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+    Mod = auto()
+    FloorDiv = auto()
+    CeilDiv = auto()
+    Constant = auto()
+    DimId = auto()
+    SymbolId = auto()
+
+    def get_token(self):
+        """Get the token corresponding to the node kind."""
+        match self:
```
[the remark this answers is not in the document]
Comment (Groverkss) [reply xdsl/affine_ir.py:20]: It's ugly to do in python3.10. In python 3.11 there is a StringEnum. I'm fine with using this as until we move to python 3.11 i guess.

File: xdsl/affine_ir.py (hunk 15 above)
Comment (Groverkss) [reply xdsl/affine_ir.py:104]: AffineExpr is not really meant to be used by anyone. I don't exactly know what to write. AffineExpr is not even Affine lol. I'll try to write something i guess.

**Pull request merged**